### ISSUE TITLE
Should we allow promotion of iterators?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#11879]
Here is an example where the `myIter` iterator is promoted, from `puzzles/deitz/puzzle_promote_iterator.chpl` :

```chpl
iter myIter(n:int) {
  for idx in 1..n do
    yield idx;
}

for thing in myIter(1..4) {
  writeln(thing);
}
```

If we allow promoted of iterators, what should this example do and what should the `thing` be?

Here is a more interesting example, inspired by `studies/lsms/shemmy/m-lsms.chpl` :

```chpl
const GridDom  = {0..#span_x, 0..#span_y, 0..#span_z};
const GridDist = GridDom dmapped Block(GridDom);

var lizDoms: [GridDist] sparse subdomain(GridDom);

forall a in lizDoms { ...... }
```

Here, the forall loop is over a Block-distributed array `lizDoms` of sparse subdomains. (Using plain default-rectangular domains, ex. `domain(1)`, would work fine, too.)

The compiler looks for a standalone iterator for `lizDoms.these()`. Since our Block distribution does not provide standalone iterators, the compiler tries promotion. It discovers that the array elements, i.e. sparse subdomains, DO define a standalone iterator.

So, instead of parallel iteration over `lizDoms` elements, the `forall` loop invokes parallel iteration over each of them. This is most likely not what the user intended. Also, currently our implementation hits bugs when compiling this example.

Therefore we could choose:

* Disable promotion of iterators and things that behave like iterators, i.e. iterator forwarders. This will make the `puzzle_promote_iterator.chpl` test invalid. Implemented in 0d241fc436. (Suggested by @benharsh.)

* Disable promotion only of `these` iterators. (Suggested by @mppf.) Implemented in ae21c9b5c3.

Both options result in correct/expected behavior of the `m-lsms.chpl`-inspired example above, by invoking the leader/follower iterators for Block instead of the standalone.

Either option would be supported by the discussion in #11847.

**Comment by bradcray** [comment]
When I posed the questions that led to this issue ("Do we need to promote iterators?  Do we rely on it?"), I wasn't sure what to expect, but was having trouble coming up with examples that would want to do it.  That said, I find Steve's puzzle intriguing because the output and behavior seem relatively intuitive and made me happy even though it took me awhile to reason through what it was doing and why.  Vass wrote:

> Here is an example where the `myIter` iterator is promoted, from `puzzles/deitz/puzzle_promote_iterator.chpl` :
> 
> ```chpl
> iter myIter(n:int) {
>   for idx in 1..n do
>     yield idx;
> }
> 
> for thing in myIter(1..5) {
>   writeln(thing);
> }
> ```
> 
> If we allow promoted of iterators, what should this example do and what should the `thing` be?

First, for those who haven't compiled and run this (or looked at the `.good` file), the output is:

```
1
1 2
1 2 3
1 2 3 4
1 2 3 4 5
```

And this is because we're effectively changing the code into something like:

```chapel
for thing in (forall j in 1..5 do let tmp = myIter(j) in tmp) {
  writeln(thing);
```

Where the key is `tmp` which is an iterable expression. writeln() prints such expressions by evaluating them and printing each of its values with space separators.

Also interesting is that the program also works if the for loop is turned into a forall loop:

```chapel
forall thing in myIter(1..5) do
  writeln(thing);
```

where the output is a (nondeterministic) permutation of the previous:

```
1 2 3 4 5
1 2 3 4
1
1 2 3
1 2
```

This seems intriguingly powerful to me, where my interpretation of why this should work is the following:

* seeing a promotion like `myIter(1..5)`, the compiler creates an iterator record that supports the same iterators that the promoting expression(s) do(es).  In this case, since `1..5` is doing the promotion—and since ranges support serial, standalone, and leader-follower iterators—the compiler would create all four iterators for the iterator record as well.

* then, applying the for-loop to the promotion invokes the serial iterator (where applying the forall-loop would invoke the standalone iterator; or leader+follower if the promoted expression didn't have a standalone iterator).

I think that this is consistent with what's being proposed in issue #11819.

This makes me think that maybe taking the more modest approach of just outlawing promotions of `these` is the right way to go (or if we wanted to be even more precise... maybe only outlaw compiler-generated calls to promotions of `these`?)

Are there downsides to such an approach that I'm not anticipating?

Also: Why the heck doesn't the Block distribution support a standalone iterator yet...?!?  :)

**Comment by mppf** [comment]
> That said, I find Steve's puzzle intriguing because the output and behavior seem relatively intuitive 

I just find it really confusing, personally. I modified it to print out the type of `thing` and the compiler indicated that `thing` is some kind of `iterator record`. I think lots of Chapel programming can make sense without resorting to a user view of iterator records. Of course, users need a mental model of what an iterator does. But iterator records are different. Maybe we want them to be user visible anyway. But I think that it's hard to explain this case without talking about iterator records and I find that concerning. (And, anytime iterator records become visible to the user, I start to worry about if they could possibly capture something by reference that then goes out of scope).

I tried some variations to possibly expose problems with iterator records capturing by reference and promotion. The 3rd variation below (testCoforall) seems to have race conditions. Is that a bug? Or a fundamental problem? I'm not sure, but I find it concerning.

``` chapel
config const m = 25;

iter myIter(const ref n:int) {
  for idx in 1..n do
    yield idx;
}

proc testFor() {
  writeln("testFor");

  var sums:[1..m] int;

  for (i,thing) in zip(1..m, myIter(1..m)) {
    sums[i] = + reduce thing;
  }

  for i in 1..m {
    assert(sums[i] == i*(i+1)/2);
  }
}
testFor();

proc testForall() {
  writeln("testForall");

  var sums:[1..m] int;

  forall (i,thing) in zip(1..m, myIter(1..m)) {
    sums[i] = + reduce thing;
  }

  for i in 1..m {
    assert(sums[i] == i*(i+1)/2);
  }
}
testForall();

proc testCoforall() {
  writeln("testCoforall");

  var sums:[1..m] int;

  coforall (i,thing) in zip(1..m, myIter(1..m)) {
    sums[i] = + reduce thing;
  }

  for i in 1..m {
    assert(sums[i] == i*(i+1)/2);
  }
}
testCoforall();
```

**Comment by bradcray** [comment]
[Disclaimer: I haven't gotten my head around Michael's examples yet]

My answer was definitely predicated on an assumption that we'll probably ultimately expose some notion of an "iterator expression" (which I think is what an iterator record represents) to the user.  Arguably we already do so today, but just don't give them the type machinery to express them and reason about them directly:

```chapel
proc foo(x) {
  writeln(x.type:string);
}

const ref r = [i in 1..10] i;
writeln(r.type:string);

foo([i in 1..10] i);
```

but I agree that if we were not to do so, supporting such promotions might not be as sensible.

I should clarify, though, that when I said that I found the output intuitive (or maybe "unsurprising" would've been a better label) that it was less about the iterator record itself, and more that it captured the notion of invoking the iterator with each of the integers in the range serially (or in parallel, with edits).

OTOH, given that we haven't seen this pattern come up much, maybe it's not that important.  But given that it seems arguably useful and compact, I wouldn't be opposed to continuing to support it for the time being.

**Comment by vasslitvinov** [comment]
I agree with Michael and Brad that we may want to expose `thing` (from the above examples) to the user as some kind of iterable. Today it is an `_iteratorRecord`, which lacks clarity and user-level typing.

Once such an iterable has been presented as a user-level concept, then iterator promotion should be unsurprising indeed.

Therefore let us continue supporting iterator promotion, except for `these`. We can allow promotion of `these` when the invocation is written by the user, rather than implicitly added by the compiler. I would like to leave this latter adjustment as future work.

**Comment by vasslitvinov** [comment]
> The 3rd variation below (testCoforall) seems to have race conditions.

I am not seeing it. (I haven't tried compiling+running it.) `testCoforall` should generate a separate instance of myIter's `_iteratorRecord` for each `i`, avoiding races. It may well be that this is not how we currently do it in the compiler.

Promotion can surely cause race conditions. I do not see when that would be specific to promotion of iterators.

> Why the heck doesn't the Block distribution support a standalone iterator yet?!?

Just to make sure that our compiler handles that case correctly?!! :)

**Comment by mppf** [comment]
>  >    The 3rd variation below (testCoforall) seems to have race conditions.

> I am not seeing it. (I haven't tried compiling+running it.)

I wasn't spotting a conceptual race condition but rather I was observing incorrect output when running it.
